pkg/cloud/libvirt/actuators/machine/actuator.go

```diff
@@ -321,7 +321,7 @@ func ProviderConfigMachine(codec codec, ms *machinev1.MachineSpec) (*providercon
 }
 
 // updateStatus updates a machine object's status.
-func (a *Actuator) updateStatus(machine *machinev1.Machine, dom *libvirt.Domain) error {
+func (a *Actuator) updateStatus(machine *machinev1.Machine, dom *libvirt.Domain, client libvirtclient.Client) error {
```

Comment:
Elsewhere we use the `clientBuilder` attached to the actuator object to build a client based on the URI in the Machine's provider config rather than passing a client around. Any reason not to do that here?

Reply:
@bison We call `updateStatus` under the `Create` and `Update` methods that already created libvirt connection and I want to avoid to create redundant libvirt connections, also it does not the single method where we pass `libvirt.Client` interface as parameter, for example we have:
- `deleteVolumeAndDomain`
- `createVolumeAndDomain`

But to be honest I do not have strong opinion here, so if it is critical I can build libvirt connection under the `updateStatus`.

Reply:
Caching the connection is IMO a very good idea. I don't see why the connection has to be created and deleted all the time. Having said that, that's a separate logical change so shouldn't it go in a separate patch?

Reply:
The assumption is every machine object can have it's own libvirt connection. E.g. One can have two libvirt services running on the same platform (e.g. one with very quickly starting guest but with minimum resources, other starting slowly but with more resources). Thus, two connections. It's then hard to use the same connection for both `deleteVolumeAndDomain` and `createVolumeAndDomain`.

Reply:
As long as the caching stores connections by a `machine.Name` and there is an upper bound on the number of open connections, I am for optimizing the code. Though, as @zeenix pointed out, that's for a different PR.

Reply:
Sharing the connection with `updateStatus` is perfectly fine though.

Reply:
Just for the record, it doesn't have to be a separate PR. Just separate commit.

Reply:
I'm not sure I understand. If the two separate connections are for different services (i-e different connection URL), you can't delete a given pair of volume and domain on a connection different from the one you created them on.

Reply:
You are right, if a volume/domain was created by a connection A, it needs to be deleted by the same connection A. I meant a scenario, where a cluster can have two machine objects. Each object with its own connection. So, when refactoring and caching libvirt connections we need to make sure a machine uses the right connection.

Reply:
> I meant a scenario, where a cluster can have two machine objects. Each object with its own connection. So, when refactoring and caching libvirt connections we need to make sure a machine uses the right connection.

This is my concern. The libvirt connection URI is part of a Machine's provider config. It's possible here to pass a client for a libvirt instance that is *not* associated with Machine object passed. It seems safer to build the client like we do elsewhere. Having said that, we should absolutely make the code that builds the connections smarter and cache open connections for each unique URI or something.

Reply:
@ingvagabund @zeenix I prefer to implement caching logic under the separate PR because it will need to implement some logic to check if the libvirt connections are alive and it does not really in the scope of this PR and can delay the review process.

Reply:
Also, let's keep in mind the libvirt url can contain a lot of parameters, e.g. ssh key. E.g. `qemu+ssh://root@${LIBVIRT_IP}/system?no_verify=1&keyfile=/root/.ssh/actuator.pem/privatekey`. So the caching needs to take into account that as well.

Reply:
Also, it might be insecure to share the connection from the security point of view.

Reply:
Isn't creating a connection object from a given URI the task for libvirt? I mean do we really need to care what the URL is? I don't think we need to verify the URL to be correctly formatted as libvirt just won't create a connection in case of a bad URL.